pandas/core/dtypes/base.py

```diff
@@ -40,11 +43,19 @@ def __eq__(self, other):
         bool
         """
         if isinstance(other, compat.string_types):
-            return other == self.name
-        elif isinstance(other, type(self)):
-            return True
-        else:
-            return False
+            try:
+                other = self.construct_from_string(other)
+            except TypeError:
+                return False
+        if isinstance(other, type(self)):
+            return all(
+                getattr(self, attr) == getattr(other, attr)
+                for attr in self._metadata
+            )
+        return False
+
+    def __hash__(self):
+        return hash(tuple(getattr(self, attr) for attr in self._metadata))
```

Comment:
shouldn't this contain something more than only the attributes? 
Because this would mean that two classes with the same attributes can give the same hash? (eg a fictional DatetimeDtype and PeriodDtype in case both have a `freq` (and only a freq))

Also, the consequence is that each attribute needs to be hashable, and eg cannot be a list?

Reply:
> Also, the consequence is that each attribute needs to be hashable, and eg cannot be a list?

Ah, I see your note about that in the docs

Reply:
> Because this would mean that two classes with the same attributes can give the same hash? 

Yes, but I don't think that's a problem in practice. As I understand it, python expects that if two classes are equal, then they have the same hash. The reverse is not true. I *think* it's OK for two objects to have the same hash, but not be equal.


With the following


```python
class A(ExtensionDtype):
    _freq = ('foo',)
    name = "A"
    def __init__(self, foo):
        self.foo = foo


class B(ExtensionDtype):
    _freq = ('foo',)
    name = 'B'

    def __init__(self, foo):
        self.foo = foo
```


We get

```python
In [2]: a = A('foo')

In [3]: b = B('foo')

In [4]: hash(a) == hash(b)
Out[4]: True

In [5]: d = {a: 'a', b: 'b'}

In [6]: d
Out[6]:
{<pandas.core.dtypes.dtypes.A at 0x11e824080>: 'a',
 <pandas.core.dtypes.dtypes.B at 0x11e80a828>: 'b'}

In [7]: d[a]
Out[7]: 'a'

In [8]: len({a, b})
Out[8]: 2
```


> Also, the consequence is that each attribute needs to be hashable, and eg cannot be a list?

Yes, that's correct. I think that's OK. If people have non-hashable things that characterize their class, they'll need to implement their own `__hash__`